File: sdk/swift/Sources/XAI/API/StakingClient.swift
import Foundation

/// Staking status.
public enum StakingStatus: String, Codable, Sendable {
    case active = "active"
    case unbonding = "unbonding"
    case inactive = "inactive"
}

/// Represents a staking position.
public struct StakingPosition: Codable, Sendable {
    /// Unique position ID
    public let id: String

    /// Staker address
    public let address: String

    /// Validator address
    public let validatorAddress: String

    /// Staked amount
    public let amount: String

    /// Staking status
    public let status: StakingStatus

    /// Start timestamp
    public let startedAt: Date

    /// Unbonding end time (if unbonding)
    public let unbondingEndsAt: Date?

    /// Accumulated rewards
    public let rewards: String

    /// Last reward claim timestamp
    public let lastClaimAt: Date?
}

/// Represents a validator.
public struct Validator: Codable, Sendable {
    /// Validator address
    public let address: String

    /// Validator name
    public let name: String

    /// Validator description
    public let description: String

    /// Commission rate (percentage)
    public let commissionRate: Double

    /// Total staked amount
    public let totalStaked: String

    /// Number of delegators
    public let delegators: Int

    /// Whether validator is active
    public let active: Bool

    /// Uptime percentage
    public let uptime: Double

    /// Voting power
    public let votingPower: String
}

/// Staking rewards information.
public struct StakingRewards: Codable, Sendable {
    /// Total rewards
    public let total: String

    /// Pending rewards (unclaimed)
    public let pending: String

    /// Claimed rewards
    public let claimed: String

    /// Estimated APY
    public let estimatedApy: Double
}

/// Stake transaction result.
public struct StakeResult: Codable, Sendable {
    /// Transaction hash
    public let txHash: String

    /// Position ID
    public let positionId: String

    /// Staked amount
    public let amount: String
}

/// Unstake transaction result.
public struct UnstakeResult: Codable, Sendable {
    /// Transaction hash
    public let txHash: String

    /// Position ID
    public let positionId: String

    /// Unstaked amount
    public let amount: String

    /// When unbonding ends
    public let unbondingEndsAt: Date
}

/// Client for staking operations.
public struct StakingClient: Sendable {
    private let httpClient: HTTPClient

    init(httpClient: HTTPClient) {
        self.httpClient = httpClient
    }

    /// Stake tokens with a validator.
    /// - Parameters:
    ///   - address: Staker address
    ///   - validatorAddress: Validator address
    ///   - amount: Amount to stake
    /// - Returns: Stake result
    /// - Throws: `XAIError` if the request fails
    public func stake(
        address: String,
        validatorAddress: String,
        amount: String
    ) async throws -> StakeResult {
        guard !address.isEmpty && !validatorAddress.isEmpty && !amount.isEmpty else {
            throw XAIError.validationError("address, validatorAddress, and amount are required")
        }

        struct StakeRequest: Encodable {
            let address: String
            let validatorAddress: String
            let amount: String
        }

        return try await httpClient.post(
            "/staking/stake",
            body: StakeRequest(address: address, validatorAddress: validatorAddress, amount: amount)
        )
    }

    /// Unstake tokens from a validator.
    /// - Parameters:
    ///   - address: Staker address
    ///   - positionId: Staking position ID
    ///   - amount: Amount to unstake (nil for full unstake)
    /// - Returns: Unstake result
    /// - Throws: `XAIError` if the request fails
    public func unstake(
        address: String,
        positionId: String,
        amount: String? = nil
    ) async throws -> UnstakeResult {
        guard !address.isEmpty && !positionId.isEmpty else {
            throw XAIError.validationError("address and positionId are required")
        }

        struct UnstakeRequest: Encodable {
            let address: String
            let positionId: String
            let amount: String?
        }

        return try await httpClient.post(
            "/staking/unstake",
            body: UnstakeRequest(address: address, positionId: positionId, amount: amount)
        )
    }

    /// Get staking positions for an address.
    /// - Parameter address: Staker address
    /// - Returns: List of staking positions
    /// - Throws: `XAIError` if the request fails
    public func getPositions(address: String) async throws -> [StakingPosition] {
        guard !address.isEmpty else {
            throw XAIError.validationError("Address is required")
        }

        struct Response: Decodable {
            let positions: [StakingPosition]
        }

        let response: Response = try await httpClient.get("/staking/positions/\(address)")
        return response.positions
    }

    /// Get staking rewards for an address.
    /// - Parameter address: Staker address
    /// - Returns: Rewards information
    /// - Throws: `XAIError` if the request fails
    public func getRewards(address: String) async throws -> StakingRewards {
        guard !address.isEmpty else {
            throw XAIError.validationError("Address is required")
        }
        return try await httpClient.get("/staking/rewards/\(address)")
    }

    /// Claim pending staking rewards.
    /// - Parameter address: Staker address
    /// - Returns: Transaction hash
    /// - Throws: `XAIError` if the request fails
    public func claimRewards(address: String) async throws -> String {
        guard !address.isEmpty else {
            throw XAIError.validationError("Address is required")
        }

        struct ClaimRequest: Encodable {
            let address: String
        }

        struct Response: Decodable {
            let txHash: String
        }

        let response: Response = try await httpClient.post(
            "/staking/claim",
            body: ClaimRequest(address: address)
        )
        return response.txHash
    }

    /// List all validators.
    /// - Parameter activeOnly: Only return active validators
    /// - Returns: List of validators
    /// - Throws: `XAIError` if the request fails
    public func listValidators(activeOnly: Bool = false) async throws -> [Validator] {
        var queryItems: [URLQueryItem]? = nil
        if activeOnly {
            queryItems = [URLQueryItem(name: "active", value: "true")]
        }

        struct Response: Decodable {
            let validators: [Validator]
        }

        let response: Response = try await httpClient.get("/staking/validators", queryItems: queryItems)
        return response.validators
    }

    /// Get a specific validator.
    /// - Parameter address: Validator address
    /// - Returns: Validator details
    /// - Throws: `XAIError` if the request fails
    public func getValidator(address: String) async throws -> Validator {
        guard !address.isEmpty else {
            throw XAIError.validationError("Address is required")
        }
        return try await httpClient.get("/staking/validators/\(address)")
    }

    /// Get estimated staking APY.
    /// - Returns: Current APY as a percentage
    /// - Throws: `XAIError` if the request fails
    public func getEstimatedApy() async throws -> Double {
        struct Response: Decodable {
            let apy: Double
        }
        let response: Response = try await httpClient.get("/staking/apy")
        return response.apy
    }
}

File: sdk/kotlin/src/main/kotlin/com/xai/sdk/api/StakingClient.kt
package com.xai.sdk.api

import com.xai.sdk.models.PaginatedResponse
import com.xai.sdk.models.StakingInfo
import com.xai.sdk.models.TxResult
import com.xai.sdk.models.Validator
import com.xai.sdk.utils.HttpClient
import com.xai.sdk.utils.ValidationException
import kotlinx.serialization.Serializable

/**
 * Client for staking operations
 *
 * Provides methods for staking, delegation, and validator management.
 *
 * Example:
 * ```kotlin
 * val client = XAIClient(baseUrl = "http://localhost:12001")
 * val info = client.staking.getStakingInfo("0x1234...")
 * val validators = client.staking.listValidators()
 * ```
 */
class StakingClient internal constructor(private val httpClient: HttpClient) {

    @Serializable
    private data class StakingInfoResponse(
        val address: String,
        val staked_amount: String,
        val rewards: String,
        val validator_address: String? = null,
        val stake_date: String? = null,
        val unlock_date: String? = null,
        val apy_rate: String = "0"
    )

    @Serializable
    private data class ValidatorResponse(
        val address: String,
        val name: String? = null,
        val total_stake: String,
        val delegator_count: Int = 0,
        val commission: String = "0",
        val status: String = "active",
        val uptime: Double = 100.0
    )

    @Serializable
    private data class ValidatorsListResponse(
        val validators: List<ValidatorResponse>,
        val total: Int,
        val limit: Int,
        val offset: Int
    )

    @Serializable
    private data class StakeRequest(
        val from: String,
        val amount: String,
        val validator_address: String? = null,
        val signature: String
    )

    @Serializable
    private data class UnstakeRequest(
        val from: String,
        val amount: String,
        val signature: String
    )

    @Serializable
    private data class ClaimRewardsRequest(
        val from: String,
        val signature: String
    )

    /**
     * Get staking information for an address
     *
     * @param address Wallet address
     * @return Staking info including staked amount and rewards
     * @throws ValidationException if address is blank
     *
     * Example:
     * ```kotlin
     * val info = client.staking.getStakingInfo("0x1234...")
     * println("Staked: ${info.stakedAmount}")
     * println("Rewards: ${info.rewards}")
     * println("APY: ${info.apyRate}%")
     * ```
     */
    suspend fun getStakingInfo(address: String): StakingInfo {
        if (address.isBlank()) {
            throw ValidationException("Address is required", "address")
        }

        val response: StakingInfoResponse = httpClient.get("/staking/$address")

        return StakingInfo(
            address = response.address,
            stakedAmount = response.staked_amount,
            rewards = response.rewards,
            validatorAddress = response.validator_address,
            stakeDate = response.stake_date,
            unlockDate = response.unlock_date,
            apyRate = response.apy_rate
        )
    }

    /**
     * List all active validators
     *
     * @param page Page number (1-indexed)
     * @param limit Number of validators per page
     * @return List of validators
     *
     * Example:
     * ```kotlin
     * val validators = client.staking.listValidators()
     * validators.forEach { v ->
     *     println("${v.name}: ${v.totalStake} (${v.uptime}% uptime)")
     * }
     * ```
     */
    suspend fun listValidators(
        page: Int = 1,
        limit: Int = 50
    ): List<Validator> {
        val actualLimit = limit.coerceIn(1, 100)
        val offset = (page - 1) * actualLimit

        val response: ValidatorsListResponse = httpClient.get(
            "/staking/validators",
            mapOf("limit" to actualLimit, "offset" to offset)
        )

        return response.validators.map { it.toValidator() }
    }

    /**
     * List validators with pagination info
     *
     * @param page Page number (1-indexed)
     * @param limit Number of validators per page
     * @return Paginated response with validators
     */
    suspend fun listValidatorsPaginated(
        page: Int = 1,
        limit: Int = 50
    ): PaginatedResponse<Validator> {
        val actualLimit = limit.coerceIn(1, 100)
        val offset = (page - 1) * actualLimit

        val response: ValidatorsListResponse = httpClient.get(
            "/staking/validators",
            mapOf("limit" to actualLimit, "offset" to offset)
        )

        return PaginatedResponse(
            data = response.validators.map { it.toValidator() },
            total = response.total,
            limit = response.limit,
            offset = response.offset
        )
    }

    /**
     * Get validator details by address
     *
     * @param address Validator address
     * @return Validator details
     * @throws ValidationException if address is blank
     *
     * Example:
     * ```kotlin
     * val validator = client.staking.getValidator("0x1234...")
     * println("Commission: ${validator.commission}%")
     * println("Delegators: ${validator.delegatorCount}")
     * ```
     */
    suspend fun getValidator(address: String): Validator {
        if (address.isBlank()) {
            throw ValidationException("Address is required", "address")
        }

        val response: ValidatorResponse = httpClient.get("/staking/validators/$address")
        return response.toValidator()
    }

    /**
     * Stake tokens
     *
     * @param from Staker address
     * @param amount Amount to stake
     * @param validatorAddress Optional validator to delegate to
     * @param signature Transaction signature
     * @return Transaction result
     * @throws ValidationException if parameters are invalid
     *
     * Example:
     * ```kotlin
     * val result = client.staking.stake(
     *     from = "0x1234...",
     *     amount = "1000",
     *     validatorAddress = "0x5678...",
     *     signature = "0xsig..."
     * )
     * println("Stake tx: ${result.hash}")
     * ```
     */
    suspend fun stake(
        from: String,
        amount: String,
        validatorAddress: String? = null,
        signature: String
    ): TxResult {
        if (from.isBlank()) {
            throw ValidationException("From address is required", "from")
        }
        if (amount.isBlank()) {
            throw ValidationException("Amount is required", "amount")
        }
        if (signature.isBlank()) {
            throw ValidationException("Signature is required", "signature")
        }

        return httpClient.post(
            "/staking/stake",
            StakeRequest(from, amount, validatorAddress, signature)
        )
    }

    /**
     * Unstake tokens
     *
     * @param from Staker address
     * @param amount Amount to unstake
     * @param signature Transaction signature
     * @return Transaction result
     * @throws ValidationException if parameters are invalid
     *
     * Example:
     * ```kotlin
     * val result = client.staking.unstake(
     *     from = "0x1234...",
     *     amount = "500",
     *     signature = "0xsig..."
     * )
     * println("Unstake tx: ${result.hash}")
     * ```
     */
    suspend fun unstake(
        from: String,
        amount: String,
        signature: String
    ): TxResult {
        if (from.isBlank()) {
            throw ValidationException("From address is required", "from")
        }
        if (amount.isBlank()) {
            throw ValidationException("Amount is required", "amount")
        }
        if (signature.isBlank()) {
            throw ValidationException("Signature is required", "signature")
        }

        return httpClient.post(
            "/staking/unstake",
            UnstakeRequest(from, amount, signature)
        )
    }

    /**
     * Claim staking rewards
     *
     * @param from Staker address
     * @param signature Transaction signature
     * @return Transaction result
     * @throws ValidationException if parameters are invalid
     *
     * Example:
     * ```kotlin
     * val info = client.staking.getStakingInfo("0x1234...")
     * if (info.rewards.toBigDecimal() > BigDecimal.ZERO) {
     *     val result = client.staking.claimRewards(
     *         from = "0x1234...",
     *         signature = "0xsig..."
     *     )
     *     println("Claimed rewards: ${result.hash}")
     * }
     * ```
     */
    suspend fun claimRewards(
        from: String,
        signature: String
    ): TxResult {
        if (from.isBlank()) {
            throw ValidationException("From address is required", "from")
        }
        if (signature.isBlank()) {
            throw ValidationException("Signature is required", "signature")
        }

        return httpClient.post(
            "/staking/claim-rewards",
            ClaimRewardsRequest(from, signature)
        )
    }

    /**
     * Get current staking APY rate
     *
     * @return Current APY rate as a string percentage
     *
     * Example:
     * ```kotlin
     * val apy = client.staking.getAPY()
     * println("Current staking APY: $apy%")
     * ```
     */
    suspend fun getAPY(): String {
        @Serializable
        data class APYResponse(val apy: String)

        val response: APYResponse = httpClient.get("/staking/apy")
        return response.apy
    }

    /**
     * Get total staked amount in the network
     *
     * @return Total staked amount
     *
     * Example:
     * ```kotlin
     * val totalStaked = client.staking.getTotalStaked()
     * println("Total network stake: $totalStaked")
     * ```
     */
    suspend fun getTotalStaked(): String {
        @Serializable
        data class TotalStakedResponse(val total_staked: String)

        val response: TotalStakedResponse = httpClient.get("/staking/total")
        return response.total_staked
    }

    private fun ValidatorResponse.toValidator() = Validator(
        address = address,
        name = name,
        totalStake = total_stake,
        delegatorCount = delegator_count,
        commission = commission,
        status = status,
        uptime = uptime
    )
}
